include/aolayer.h

```diff
@@ -142,6 +127,10 @@ class AOLayer : public QLabel {
   // this is only initialized if network_strings has size more than 0.
   void load_network_effects();
 
+  // Play a frame-specific effect, if there's any defined for that specific
+  // frame.
+  void play_frame_effect(int frame);
+
```

Comment:
These functions are only used by the character layer. You should move them downward along with their respective fields.

Reply:
i attempted to do this but ran into an issue - this function is called from `play()`, which is used by every layer. i was able to prevent crashing by adding a sanity check (is the effects vector empty? if it is, don't do anything) but I'm not really sure how to have this function be under `CharLayer` only

Reply:
In `CharLayer`, override `play()` so it calls the superclass `play()` and then `play_frame_effect()`.